src/mame/machine/dmac3.cpp

```diff
@@ -0,0 +1,231 @@
+	m_dma_w.resolve_all_safe();
+
+	m_irq_check = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(dmac3_device::irq_check), this));
+	m_dma_check = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(dmac3_device::dma_check), this));
+}
+
+uint32_t dmac3_device::csr_r(dmac3_controller controller)
+{
+	return m_controllers[controller].csr;
+}
+
+uint32_t dmac3_device::intr_r(dmac3_controller controller)
+{
+	return m_controllers[controller].intr;
+}
+
+uint32_t dmac3_device::length_r(dmac3_controller controller)
+{
+	return m_controllers[controller].length;
+}
+
+uint32_t dmac3_device::address_r(dmac3_controller controller)
+{
+	return m_controllers[controller].address;
+}
+
+uint32_t dmac3_device::conf_r(dmac3_controller controller)
+{
+	return m_controllers[controller].conf;
+}
+
+void dmac3_device::csr_w(dmac3_controller controller, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTER, "dmac3-%d csr_w: 0x%x\n", controller, data);
+	if (data & CSR_RESET)
+	{
+		LOGMASKED(LOG_GENERAL, "dmac3-%d chip reset\n", controller);
+		reset_controller(controller);
+	}
+	else
+	{
+		m_controllers[controller].csr = data;
+	}
+}
+
+void dmac3_device::intr_w(dmac3_controller controller, uint32_t data)
+{
+	static constexpr uint32_t INTR_CLR_MASK = (INTR_INT | INTR_TCI | INTR_EOP | INTR_EOPI | INTR_DREQ | INTR_DRQI | INTR_PERR);
+	static constexpr uint32_t INTR_EN_MASK = (INTR_INTEN | INTR_TCIE | INTR_EOPIE | INTR_DRQIE);
+
+	LOGMASKED(LOG_REGISTER, "dmac3-%d intr_w: 0x%x\n", controller, data);
+	const auto intr_clear_bits = data & INTR_CLR_MASK;  // Get 1s on bits to clear
+	const auto intr_enable_bits = data & INTR_EN_MASK;  // Get 1s on bits to set
+	m_controllers[controller].intr &= ~intr_clear_bits; // Clear requested interrupt flags
+	m_controllers[controller].intr &= ~INTR_EN_MASK;    // Clear all mask bits
+	m_controllers[controller].intr |= intr_enable_bits; // Set mask bits to new mask
+	m_irq_check->adjust(attotime::zero);
+}
+
+void dmac3_device::length_w(dmac3_controller controller, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTER, "dmac3-%d length_w: 0x%x (%s)\n", controller, data, machine().describe_context());
+	m_controllers[controller].length = data;
+}
+
+void dmac3_device::address_w(dmac3_controller controller, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTER, "dmac3-%d address_w: 0x%x (%s)\n", controller, data, machine().describe_context());
+	m_controllers[controller].address = data;
+}
+
+void dmac3_device::conf_w(dmac3_controller controller, uint32_t data)
+{
+#if (VERBOSE & LOG_REGISTER) > 0 
+	// No need for the extra comparison if logging isn't enabled
+	// Only log if something other than the access mode changed
+	// since, at least for now, DMAC and SPIFI accesses will go through regardless
+	// of this setting.
+
+	if ((data & ~CONF_WIDTH) != (m_controllers[controller].conf & ~CONF_WIDTH))
+	{
+		LOGMASKED(LOG_REGISTER, "dmac3-%d conf_w: 0x%x (%s)\n", controller, data, machine().describe_context());
+	}
+#endif
```

Comment:
There’s no point putting the `#if` around this – the compiler will optimise away the test if logging is disabled because it has no side effects and doesn’t generate a sequence point.  The `#if` is harmful because it means the code inside it can rot if people don’t go out of their way to test with logging enabled.